winch/codegen/src/isa/x64/asm.rs

```diff
@@ -472,6 +513,46 @@ impl Assembler {
         });
     }
 
+    /// Compare two operands and write bit to dst
+    pub fn cmp(&mut self, src: Operand, dst: Operand, kind: CmpKind, size: OperandSize) {
+        let src = GprMemImm::new(match src {
+            Operand::Reg(r) => r.into(),
+            Operand::Imm(imm) => match i32::try_from(imm) {
+                Ok(i) => RegMemImm::imm(i as u32),
+                Err(_) => {
+                    let scratch = regs::scratch();
+                    self.mov_ir(imm as u64, scratch, size);
+                    scratch.into()
+                }
+            },
+            Operand::Mem(_) => panic!("mem operand for src not supported"),
+        })
+        .expect("valid register or immediate");
+        let dst = match dst {
+            Operand::Reg(r) => r,
+            _ => panic!("dst must be a register"),
+        };
```

Comment:
I'm open to suggestions on better ways to get the `src` and `dst` arguments from `Operands` into the correct types.

If we do opt for separate `cmp_rr` and `cmp_ir` methods, I think I would want to figure out some way to re-use the implementation of the three emits across both implementations given that it is three shared instructions so I don't think it should be inlined in both places.

Reply:
I think that having `cmp_rr` and `cmp_ir` is what I'd prefer, to keep the lowering symmetric with the other functions, but also because doing so leaves the door open to easily add lowering for other types of operands (e.g. the memory case).

As far as sharing the emission of the three instructions I think that a good option here would be to:

* Move the emission of `setcc` into its own function in the assembler, as described in https://github.com/jeffcharles/wasmtime/pull/3/files#r1210412335
* Avoid emitting `setcc` in the assembler's compare instruction and instead move the emission of `setcc` at the masm level, into a method that explicitly communicates that the `setcc` will be emitted:


```rs
// masm.rs

/// Perform comparison with set.
pub fn cmp_set(...) {
  self.asm.cmp();
  self.asm.setcc();
}

```

This approach has the benefit that doesn't tie the `cmp` instruction at the `asm` level to emit `setcc`, which won't be required in all cases, like for example, in trampolines you might want to branch directly instead of emitting `setcc`, or when optimizing control flow, so it will give the freedom to the caller to either call `cmp_set` or `cmp` depending on their use-case. 


Once we need to branch in trampolines or once we optimize control flow my expectation is that we'd introduce `cmp` and perpahs `branch` methods at the masm level:


```rs

pub fn cmp(...) {}
pub fn branch( ...) {
  self.cmp();
  self.asm.br(...)
}
```